src/icon.c

```diff
@@ -15,38 +16,96 @@ static bool is_readable_file(const char *filename)
         return (access(filename, R_OK) != -1);
 }
 
-static cairo_status_t read_from_buf(void *closure, unsigned char *data, unsigned int size)
+/**
+ * Reassemble the data parts of a GdkPixbuf into a cairo_surface_t's data field.
+ *
+ * Requires to call on the surface flush before and mark_dirty after the execution.
+ */
+static void pixbuf_data_to_cairo_data(
+                const unsigned char *pixels_p,
+                      unsigned char *pixels_c,
+                size_t rowstride_p,
+                size_t rowstride_c,
+                int width,
+                int height,
+                int n_channels)
 {
-        GByteArray *buf = (GByteArray *)closure;
-
-        unsigned int cpy = MIN(size, buf->len);
-        memcpy(data, buf->data, cpy);
-        g_byte_array_remove_range(buf, 0, cpy);
+#if G_BYTE_ORDER == G_LITTLE_ENDIAN
+        static const size_t CAIRO_B = 0;
+        static const size_t CAIRO_G = 1;
+        static const size_t CAIRO_R = 2;
+        static const size_t CAIRO_A = 3;
+#elif G_BYTE_ORDER == G_BIG_ENDIAN
+        static const size_t CAIRO_A = 0;
+        static const size_t CAIRO_R = 1;
+        static const size_t CAIRO_G = 2;
+        static const size_t CAIRO_B = 3;
+#elif G_BYTE_ORDER == G_PDP_ENDIAN
+        static const size_t CAIRO_R = 0;
+        static const size_t CAIRO_A = 1;
+        static const size_t CAIRO_B = 2;
+        static const size_t CAIRO_G = 3;
+#else
+// GLib doesn't support any other endiannesses
+#error Unsupported Endianness
+#endif
+
+        assert(pixels_p);
+        assert(pixels_c);
+        assert(width > 0);
+        assert(height > 0);
+
+        if (n_channels == 3) {
+                for (int h = 0; h < height; h++) {
+                              unsigned char *iter_c = pixels_c + h * rowstride_c;
+                        const unsigned char *iter_p = pixels_p + h * rowstride_p;
+                        for (int w = 0; w < width; w++) {
+                                iter_c[CAIRO_R] = iter_p[0];
+                                iter_c[CAIRO_G] = iter_p[1];
+                                iter_c[CAIRO_B] = iter_p[2];
+                                iter_c[CAIRO_A] = 0xff;
+                                iter_c += 4;
+                                iter_p += n_channels;
+                        }
+                }
+        } else {
+                for (int h = 0; h < height; h++) {
+                              unsigned char *iter_c = pixels_c + h * rowstride_c;
+                        const unsigned char *iter_p = pixels_p + h * rowstride_p;
+                        for (int w = 0; w < width; w++) {
+                                double alpha_factor = iter_p[3] / (double)0xff;
+                                iter_c[CAIRO_R] = (unsigned char)(iter_p[0] * alpha_factor + .5);
```

Comment:
Can you explain `alpha_factor` and why you add `.5` here?

Reply:
Docs are here:

https://cairographics.org/manual/cairo-Image-Surfaces.html#cairo-format-t

I've actually taken [this gist](https://gist.github.com/bert/985903/1f7675464104c70fb65ed93328b7538f459775dc#file-draw_in_a_gdkpixbuf_with_cairo_and_pango-c-L104-L210) in my first iteration.